design/20240625.push-charts-to-oci.md

```diff
@@ -0,0 +1,251 @@
+<!--
+This template is adapted from Kubernetes Enhancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
+-->
+
+# Push cert-manager Helm Charts to an OCI Registry
+
+<!-- toc -->
+- [Release Signoff Checklist](#release-signoff-checklist)
+- [Summary](#summary)
+- [Motivation](#motivation)
+  - [Goals](#goals)
+  - [Non-Goals](#non-goals)
+- [Proposal](#proposal)
+  - [Risks and Mitigations](#risks-and-mitigations)
+- [Design Details](#design-details)
+  - [Test Plan](#test-plan)
+  - [Graduation Criteria](#graduation-criteria)
+  - [Upgrade / Downgrade Strategy](#upgrade--downgrade-strategy)
+  - [Supported Versions](#supported-versions)
+- [Production Readiness](#production-readiness)
+- [Drawbacks](#drawbacks)
+- [Alternatives](#alternatives)
+<!-- /toc -->
+
+## Release Signoff Checklist
+
+This checklist contains actions which must be completed before a PR implementing this design can be merged:
+
+- [ ] This design doc has been discussed and approved
+- [ ] Test plan has been agreed upon and the tests implemented
+- [ ] Feature gate status has been agreed upon (whether the new functionality will be placed behind a feature gate or not)
+- [ ] Graduation criteria is in place if required (if the new functionality is placed behind a feature gate, how will it graduate between stages)
+- [ ] User-facing documentation has been PR-ed against the release branch in [cert-manager/website]
+
+## Summary
+
+<!--
+This section is important for producing high-quality, user-focused
+documentation such as release notes.
+
+A good summary is probably around a paragraph in length.
+
+[documentation style guide]: https://github.com/kubernetes/community/blob/master/contributors/guide/style-guide.md
+-->
+
+This design proposes to start pushing cert-manager's Helm charts to an OCI registry, `quay.io/cert-manager` beginning with cert-manager 1.16.
+
+It proposes no other changes - specifically, after this design is implemented charts would still be pushed to their current location for every
```

Comment:
I think it would make sense for each project to only publish to the new location and for a backwards-compatibility script to copy the charts and images to the old locations.
Additionally, we could rewrite the referenced images in the old charts location to keep pointing to the old images.
We can also keep some other compatibility hacks in this separate repo (eg. adding/removing the v prefix from charts), without polluting each individual project.

Another alternative: We could also make all the individual projects push to a "staging" environment and use that location as the source of truth, also for the "official" download location. We could then also run additional sanity checks and additional signatures/ image rewrites before we publish these charts to the "official" download location.

Reply:
> I think it would make sense for each project to only publish to the new location and for a backwards-compatibility script to copy the charts and images to the old locations.

I think mirroring like this is definitely valid and it would make sense for us to do it this way as long as we're running the mirroring script during the release and not just relying on a cron trigger or something. I don't think we should consider the release to be complete until all artifacts are available in all locations.

I'm not sure about making any changes to the charts as part of the release process though. If we're saying that charts in `quay.io/jetstack` should differ from `quay.io/cert-manager` I can understand that argument as long as the charts for a project can be built from a checkout of the project.

I think having a staging location would be an _excellent_ improvement and would be fully in favour of that.

---

All of that said, I feel like a lot of that is a big increase of scope for this design. All I'm really trying to do here is solve pushing helm charts to an OCI registry in the short term.

It's a fair point that we might want charts in `quay.io/cert-manager` to use images from `quay.io/cert-manager`, and that might be a point for further discussion.

That makes me wonder: **What if** this design was reworked to push charts to `quay.io/jetstack` instead? That leaves all options on the table when we migrate away from the jetstack name and gives us time for those discussions.